src/virtual_memory/anonymous.cpp

```diff
@@ -0,0 +1,111 @@
+/**
+ * @file
+ * @brief This file implements the virtual memory and virtual address handling
+ * @copyright Eta Scale AB. Licensed under the Eta Scale Open Source License. See the LICENSE file for details.
+ */
+
+#include <cerrno>
+#include <cstddef>
+#include <fcntl.h>
+#include <iostream>
+#include <string>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/statvfs.h>
+#include <sys/types.h>
+#include <system_error>
+#include <unistd.h>
+#include "virtual_memory.hpp"
+
+/** @todo hardcoded start address */
+static const char* ARGO_START = (char*) 0x200000000000l;
+/** @todo hardcoded end address */
+static const char* ARGO_END   = (char*) 0x600000000000l;
+/** @todo hardcoded size */
+static const ptrdiff_t ARGO_SIZE = ARGO_END - ARGO_START;
+
+/** @brief the address at which the backing storage space starts */
+static char* backing_addr;
+
+/** @brief the current offset into the backing storage space */
+static std::size_t backing_offset;
+
+/** @brief the offset within the memory region controlled to the beginning of the virtual address space */
+static std::size_t file_offset;
+
+/** @brief error message string */
+std::string msg_insufficient_memory = "ArgoDSM anonymous mappable memory is insufficient.";
+/** @brief error message string */
+std::string msg_mprotect_fail = "ArgoDSM could not mprotect anonymous mapping";
+/** @brief error message string */
+std::string msg_invalid_remap = "ArgoDSM encountered an invalid mapping attempt in virtual address space";
+/** @brief error message string */
+std::string msg_main_mmap_fail = "ArgoDSM failed to set up virtual memory. Please report a bug.";
+
+namespace argo {
+	namespace virtual_memory {
+		void init() {
+			/** @todo check desired range is free */
+			backing_addr = static_cast<char*>(
+				::mmap((void *)ARGO_START, ARGO_SIZE, PROT_NONE, MAP_ANONYMOUS|MAP_SHARED|MAP_FIXED, -1, 0)
+				);
+			if(backing_addr == MAP_FAILED) {
+				std::cerr << msg_main_mmap_fail << std::endl;
+				throw std::system_error(std::make_error_code(static_cast<std::errc>(errno)), msg_main_mmap_fail);
+				exit(EXIT_FAILURE);
+			}
+			char* virtual_addr = (char*) 0x300000000000l;
+			backing_offset = 0;
+			file_offset = virtual_addr - backing_addr;
+		}
+
+		void* start_address() {
+			return backing_addr + file_offset;
+		}
+
+		std::size_t size() {
+			return ARGO_SIZE/4;
+		}
+
+		void* allocate_mappable(std::size_t alignment, std::size_t size) {
+			/* compute next free well-aligned offset */
+			backing_offset = ((backing_offset + alignment - 1)/alignment);
+			backing_offset *= alignment;
```

Comment:
this function has a failure condition (returns nullptr if not enough space is available), which is not clear to me what we should feed into it.